pkg/pillar/cmd/zedrouter/patchenvelopes.go

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) 2023 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package zedrouter
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/lf-edge/eve/pkg/pillar/base"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	fileutils "github.com/lf-edge/eve/pkg/pillar/utils/file"
+)
+
+// PatchEnvelopes is a structure representing
+// Patch Envelopes exposed to App instances via metadata server
+// for more info check docs/PATCH-ENVELOPES.md
+// Must be created by calling NewPatchEnvelopes()
+//
+// Internally, PatchEnvelopes structure stores envelopes which
+// come from EdgeDevConfig parsed by zedagent. This envelopes contains
+// both inline binary artifacts which are ready to be downloaded by app instances
+// and volume references, which are handled by volumemgr. So PatchEnvelopes struct has
+// VolumeStatusCh and PatchEnvelopeInfoCh to process volumes and patch envelopes.
+// NewPatchEnvelopes() starts coroutine processMessages which reads from both channels
+// and stores latest state of patch envelopes exposed to app instances via Get()
+//
+// That means that VolumeStatusCh and PatchEnvelopeInfoCh should not be read from anywhere
+// except processMessages(). so there is no deadlocks.
```

Comment:
Please add that it also means that processMessages can not wait/hang (e.g., by sending on a channel) since that could also lead to a deadlock.

Reply:
added